### PR TITLE
Parameter Sweep Classs

### PR DESCRIPTION
**Pull request by k1nshuk**
## Fixes/Resolves:

Reorganizes parameter sweep functions into a class to enable different type of analyses

## Summary/Motivation:

The ability to create parameter sweep objects facilitates different types of analyses. Additionally, this will make it more aligned with OOP instead of functional programming.

## Changes proposed in this PR:
- Create two different classes for Parameter Sweep and Recursive Parameter Sweep
- Update all calls to the parameter sweep in the remainder of the code.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/tools/parameter_sweep/parameter_sweep_class.py [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+
+        Returns:
+
+            save_data : A list were the first N columns are the values of the parameters passed
+                        by ``sweep_params`` and the remaining columns are the values of the
+                        simulation identified by the ``outputs`` argument.
+        """
+
+        # Initialize the base Class
+        _ParameterSweepBase.__init__(self)
```
Comment (bknueven) [inline watertap/tools/parameter_sweep/parameter_sweep_class.py]:
```suggestion
        super().__init__()
```

File: watertap/tools/parameter_sweep/parameter_sweep_class.py [hunk 2]
```diff
@@ -0,0 +1,1066 @@
...
+        #         self.write_outputs = False
+        #     else:
+        #         self.write_outputs = True
+        self.writer = ParameterSweepWriter(
+            self.comm,
+            csv_results_file_name=csv_results_file_name,
+            h5_results_file_name=h5_results_file_name,
+            debugging_data_dir=debugging_data_dir,
+            interpolate_nan_outputs=interpolate_nan_outputs,
+        )
```
Comment (bknueven) [inline watertap/tools/parameter_sweep/parameter_sweep_class.py]: It looks to me like both can go in `_ParameterSweepBase.__init__`

File: watertap/tools/parameter_sweep/parameter_sweep_class.py [hunk 3]
```diff
@@ -0,0 +1,1066 @@
...
+
+        local_output_dict["solve_successful"] = local_solve_successful_list
+
+        return local_output_dict
+
+    # ================================================================
+
+    @abstractmethod
+    def parameter_sweep(self, *args, **kwargs):
+        raise NotImplementedError
```
Comment (bknueven) [inline watertap/tools/parameter_sweep/parameter_sweep_class.py]:
```suggestion
        pass
```

File: watertap/examples/flowsheets/RO_with_energy_recovery/monte_carlo_sampling_RO_ERD.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 ###############################################################################
 
 from idaes.core.solvers import get_solver
-from watertap.tools.parameter_sweep import (
+from watertap.tools.parameter_sweep.parameter_sweep import (
```
Comment (bknueven) [inline watertap/examples/flowsheets/RO_with_energy_recovery/monte_carlo_sampling_RO_ERD.py]: This is okay, but we should put the appropriate public objects and functions in `watertap/tools/paramter_sweep/__init__.py`

File: watertap/tools/parameter_sweep/parameter_sweep_class.py (hunk 2 above)
Comment (k1nshuk) [reply watertap/tools/parameter_sweep/parameter_sweep_class.py]: Moved into a base class in a more recent commit.

**Comment by bknueven** [comment]
@k1nshuk something I noticed while I was preparing this for merge is that the recursive parameter sweep baseline results have changed.

Is this to be expected (e.g., does this PR fix a bug)?

**Comment by bknueven** [comment]
> @k1nshuk something I noticed while I was preparing this for merge is that the recursive parameter sweep baseline results have changed.
> 
> Is this to be expected (e.g., does this PR fix a bug)?

I talked with @k1nshuk offline and confirmed this was a bug.

**Review by bknueven — APPROVED** [review]

**Comment by bknueven** [comment]
There's more to be done here, but let's save further improvements for subsequent PRs.

File: watertap/tools/MPI/dummy_mpi.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class DummyCOMM:
+
+    rank = 0
+    size = 1
+
+    @staticmethod
+    def Get_rank():
+        return 0
```
Comment (eyoung55) [inline watertap/tools/MPI/dummy_mpi.py:23]: This is a nice way to avoid the problems of a missing mpi4py down the road.

**Review by eyoung55 — APPROVED** [review]

**Pull request merged**